compiler-core/templates/prelude.mjs

```diff
@@ -91,52 +91,253 @@ export class NonEmpty extends List {
+   *
+   * When returning the final byte of a bit array with a bit size that's not a
+   * multiple of 8, the content of the unused low bits are undefined.
+   *
+   * @param {number} index
+   * @returns {number | undefined}
+   */
   byteAt(index) {
-    return this.buffer[index];
+    if (index < 0 || index >= this.byteSize) {
+      return undefined;
+    }
+
+    return bitArrayByteAt(this.rawBuffer, this.bitOffset, index);
   }
 
-  // @internal
-  floatFromSlice(start, end, isBigEndian) {
-    return byteArrayToFloat(this.buffer, start, end, isBigEndian);
+  /** @internal */
+  equals(other) {
+    if (this.bitSize !== other.bitSize) {
+      return false;
+    }
+
+    const wholeByteCount = Math.trunc(this.bitSize / 8);
+
+    // If both bit offsets are zero do a byte-aligned equality check which is
+    // faster
+    if (this.bitOffset === 0 && other.bitOffset === 0) {
+      // Compare any whole bytes
+      for (let i = 0; i < wholeByteCount; i++) {
+        if (this.rawBuffer[i] !== other.rawBuffer[i]) {
+          return false;
+        }
+      }
+
+      // Compare any trailing bits, excluding unused low bits
+      const trailingBitsCount = this.bitSize % 8;
+      if (trailingBitsCount) {
+        const unusedLowBitCount = 8 - trailingBitsCount;
+        if (
+          this.rawBuffer[wholeByteCount] >> unusedLowBitCount !==
+          other.rawBuffer[wholeByteCount] >> unusedLowBitCount
+        ) {
+          return false;
+        }
+      }
+    } else {
+      // Compare any whole bytes
+      for (let i = 0; i < wholeByteCount; i++) {
+        const a = bitArrayByteAt(this.rawBuffer, this.bitOffset, i);
+        const b = bitArrayByteAt(other.rawBuffer, other.bitOffset, i);
+
+        if (a !== b) {
+          return false;
+        }
+      }
+
+      // Compare any trailing bits
+      const trailingBitsCount = this.bitSize % 8;
+      if (trailingBitsCount) {
+        const a = bitArrayByteAt(
+          this.rawBuffer,
+          this.bitOffset,
+          wholeByteCount,
+        );
+        const b = bitArrayByteAt(
+          other.rawBuffer,
+          other.bitOffset,
+          wholeByteCount,
+        );
+
+        const unusedLowBitCount = 8 - trailingBitsCount;
+        if (a >> unusedLowBitCount !== b >> unusedLowBitCount) {
+          return false;
+        }
+      }
+    }
+
+    return true;
   }
 
-  // @internal
-  intFromSlice(start, end, isBigEndian, isSigned) {
-    return byteArrayToInt(this.buffer, start, end, isBigEndian, isSigned);
+  /** @internal */
+  slice(start, end) {
+    return bitArraySlice(this, start, end);
   }
 
-  // @internal
-  binaryFromSlice(start, end) {
-    const buffer = new Uint8Array(
-      this.buffer.buffer,
-      this.buffer.byteOffset + start,
-      end - start
+  /** @internal */
+  sliceToFloat(start, end, isBigEndian) {
+    return bitArraySliceToFloat(this, start, end, isBigEndian);
+  }
+
+  /** @internal */
+  sliceToInt(start, end, isBigEndian, isSigned) {
```

Comment:
These methods mean that the large functions will never be removed by a JS bundler, isn't that right? Given their large size we want it to be that when unused there's no references to these functions so they get removed during bundling and minification so Gleam users don't need to pay a price for a feature they are not using.

Reply:
255e24e7b4f01be8059fd2a020e704be240f081c restructures things such that the three slicing functions are removable by esbuild. It's an unrelated change to the rest of this PR as the previous slicing functions also weren't able to be removed, but yes they've gotten larger so it matters a bit more now.

There's still an increase in JS code size due `toBitArray()` now being more complicated, the need for a `BitArray.equals()` function, the new deprecation warnings on access of `BitArray.buffer` and `BitArray.length`, etc. This code:

```gleam
pub fn main() {
  <<>>
}
```

Now bundles and minifies to 2965 bytes (1095 bytes gzip'ed) whereas previously it was 1560 bytes (676 bytes gzip'ed).

Reply:
You're the best 💜